scapy/contrib/mqttsn.py

```diff
@@ -0,0 +1,433 @@
+class MQTTSNPubrec(Packet):
+    name = "MQTT-SN publish received"
+    fields_desc = [
+        ShortField("mid", 0),
+    ]
+
+
+class MQTTSNPubrel(Packet):
+    name = "MQTT-SN publish release"
+    fields_desc = [
+        ShortField("mid", 0),
+    ]
+
+
+class MQTTSNSubscribe(Packet):
+    name = "MQTT-SN subscribe request"
+    fields_desc = FLAG_FIELDS + [
+        ShortField("mid", 0),
+        ConditionalField(ShortField("tid", 0),
+                         lambda pkt: pkt.tid_type == 0b01),
+        ConditionalField(StrFixedLenField("short_topic", 0, length=2),
+                         lambda pkt: pkt.tid_type == 0b10),
+        ConditionalField(StrField("topic", 0),
+                         lambda pkt: pkt.tid_type not in [0b01, 0b10]),
+    ]
+
+
+class MQTTSNSuback(Packet):
+    name = "MQTT-SN subscribe ACK"
+    fields_desc = FLAG_FIELDS + [
+        ShortField("tid", 0),
+        ShortField("mid", 0),
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNUnsubscribe(Packet):
+    name = "MQTT-SN unsubscribe request"
+    fields_desc = FLAG_FIELDS + [
+        ShortField("mid", 0),
+        ConditionalField(ShortField("tid", 0),
+                         lambda pkt: pkt.tid_type == 0b01),
+        ConditionalField(StrFixedLenField("short_topic", 0, length=2),
+                         lambda pkt: pkt.tid_type == 0b10),
+        ConditionalField(StrField("topic", 0),
+                         lambda pkt: pkt.tid_type not in [0b01, 0b10]),
+    ]
+
+
+class MQTTSNUnsuback(Packet):
+    name = "MQTT-SN unsubscribe ACK"
+    fields_desc = [
+        ShortField("mid", 0),
+    ]
+
+
+class MQTTSNPingReq(Packet):
+    name = "MQTT-SN ping request"
+    fields_desc = [
+        StrField("client_id", ""),
+    ]
+
+
+class MQTTSNPingResp(Packet):
+    name = "MQTT-SN ping response"
+
+
+class MQTTSNDisconnect(Packet):
+    name = "MQTT-SN disconnect request"
+    fields_desc = [
+        ConditionalField(
+            ShortField("duration", 0),
+            lambda pkt: hasattr(pkt.underlayer, "len") and
+            pkt.underlayer.len > 2
+        ),
+    ]
+
+
+class MQTTSNWillTopicUpd(Packet):
+    name = "MQTT-SN will topic update"
+    fields_desc = FLAG_FIELDS + [
+        StrField("will_topic", ""),
+    ]
+
+
+class MQTTSNWillTopicResp(Packet):
+    name = "MQTT-SN will topic response"
+    fields_desc = [
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNWillMsgUpd(Packet):
+    name = "MQTT-SN will message update"
+    fields_desc = [
+        StrField("will_msg", "")
+    ]
+
+
+class MQTTSNWillMsgResp(Packet):
+    name = "MQTT-SN will message response"
+    fields_desc = [
+        ByteEnumField("return_code", ACCEPTED, RETURN_CODES),
+    ]
+
+
+class MQTTSNEncaps(Packet):
+    name = "MQTT-SN encapsulated message"
+    fields_desc = [
+        BitField("resvd", 0, 6),
+        BitField("radius", 0, 2),
+        StrLenField(
+            "w_node_id", "",
```

Comment:
This field also should manipulate the `adjust` of the `len` field of the `MQTTSN` wrapping packet, as the `len` for this packet type only signifies the length of the MQTTSNEncaps header without the encapsulated message (see section 5.5 of the [spec](http://www.mqtt.org/new/wp-content/uploads/2009/06/MQTT-SN_spec_v1.2.pdf)). Is there a way to express this?

Reply:
Can't you use `pkt.underlayer` in the adjust ?

Reply:
In the adjust of this field you mean?

Reply:
If I see this correctly, `StrLenField` does not have an adjust parameter in its constratuctor.

Reply:
If I understand correctly, I meant
> the adjust of the len field of the MQTTSN wrapping packet

It might be better if you clarify :smile: from what understand `len` in MQTTSN needs info from `w_node_id`. It can get this info during the adjust of the `len`

Reply:
> It might be better if you clarify :smile:

Ok, I'll try my best :sweat_smile::

For all packet types except this one the `len` field of `MQTTSN` (the underlayer of these packet types) denotes the total length of the packet. Only for this it denotes the length until the end of `w_node_id` (called "wireless node id" in the spec), and the length of the remainder is denoted by `len` field of a following `MQTTSN` packet (the encapsulated MQTT-SN datagram). So only for this I need to manipulate the `len` field of `MQTTSN` somehow to only go to the end of `w_node_id` string.

> from what understand len in MQTTSN needs info from w_node_id. It can get this info during the adjust of the len

Do you mean something like this?

```py
class MQTTSN(Packet):
    name = "MQTT-SN header"
    fields_desc = [
        # …
        VariableFieldLenField("len", None, length_of="len",
                              adjust=lambda pkt, x: x + len(pkt.payload))
                              if type(pkt.payload) is not MQTTSNEncaps
                              else x + len(pkt.payload.w_node_id) + 1,
        # …
    ]
```

Reply:
Ok, I somewhat did it like that :-)